arch/x64/arch-elf.hh

```diff
@@ -42,4 +42,37 @@ enum {
 
 #define ELF_KERNEL_MACHINE_TYPE 62
 
+#define SAFETY_BUFFER 256
+#include <osv/align.hh>
+
+inline void elf_entry_point(void* ep, int argc, char** argv, int argv_size)
+{
+    //The layout of the stack and state of all relevant registers is described
+    //in detail in the section 3.4 (Process Initialization) of the System V Application
+    //Binary Interface AMD64 Architecture Processor Supplement Draft Version 0.95
+    //(see https://refspecs.linuxfoundation.org/elf/x86_64-abi-0.95.pdf)
+    int argc_plus_argv_stack_size = argv_size + 1;
+
+    void *stack;
+    asm volatile ("movq %%rsp, %0" : "=r"(stack));
+    //
+    //According to the document above the stack pointer should be 16-bytes aligned
+    stack -= (SAFETY_BUFFER + argc_plus_argv_stack_size * sizeof(char*));
+    stack = align_down(stack, 16);
```

Comment:
What is this "SAFETY BUFFER" thing? is this the red zone? The "red zone" is only 128 bytes, not 256.
Check out arch/x64/signal.cc:
```c++
    if (!rsp) {
        rsp = reinterpret_cast<void*>(ef->rsp);
        rsp -= 128;                 // skip red zone
    }
    rsp -= sizeof(signal_frame);
    // the Linux x86_64 calling conventions want 16-byte aligned rsp.
    // signal_frame may want even stricter alignment (but probably won't).
    rsp = align_down(rsp, std::max(16UL, alignof(signal_frame)));
```

Reply:
Addressed by better comments in the latest version of this PR.